### PR TITLE
MM-37775 - New run permissions

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- Moving to a permissions model where the decision steps are: 
  1. Are you an admin? If so, you can see that run info.
  2. Can you view the playbook the run is based on? (either the original playbook is public to the team and you're a team member, or the playbook is private and you're a member of the playbook) If so, you can view the run.
  3. Are you a guest? If so, you can see that run info only if you a channel member.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-37775

#### Checklist
- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [x] Unit tests updated

### CODE REVIEW

File: server/api/playbooks.go [hunk 1]
```diff
@@ -274,14 +274,14 @@ func (h *PlaybookHandler) deletePlaybook(w http.ResponseWriter, r *http.Request)
 	vars := mux.Vars(r)
 	userID := r.Header.Get("Mattermost-User-ID")
 
-	playbookToDelete, err := h.playbookService.Get(vars["id"])
-	if err != nil {
-		h.HandleError(w, err)
+	if err2 := app.PlaybookAccess(userID, vars["id"], h.playbookService, h.pluginAPI); err2 != nil {
```
Comment (lieut-data) [inline server/api/playbooks.go:277]: nit: can we avoid `err2`?

File: server/api/playbooks.go [hunk 2]
```diff
@@ -133,14 +133,14 @@ func (h *PlaybookHandler) getPlaybook(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	userID := r.Header.Get("Mattermost-User-ID")
 
-	playbook, err := h.playbookService.Get(vars["id"])
-	if err != nil {
-		h.HandleError(w, err)
+	if err := app.PlaybookAccess(userID, vars["id"], h.playbookService, h.pluginAPI); err != nil {
```
Comment (lieut-data) [inline server/api/playbooks.go:136]: nit: worth declaring `playbookID := vars["id"]` for clarity?

File: server/app/permissions.go [hunk 3]
```diff
@@ -24,9 +24,26 @@ type RequesterInfo struct {
 	IsGuest bool
 }
 
+func ViewPlaybookRunFromPlaybookRunID(userID, playbookRunID string, playbookService PlaybookService, playbookRunService PlaybookRunService, pluginAPI *pluginapi.Client) error {
```
Comment (lieut-data) [inline server/app/permissions.go:27]: nit: `UserCanViewPlaybookRun`? (Also, accepting the `id` feels like the default case, vs. something like `...FromChannelId`).

File: server/app/permissions.go [hunk 4]
```diff
@@ -24,9 +24,26 @@ type RequesterInfo struct {
...
+		return nil
+	}
+
+	return PlaybookAccess(userID, playbookRunID, playbookService, pluginAPI)
+}
+
 // ViewPlaybookRunFromChannelID returns nil if the userID has permissions to view the playbook run
 // associated with channelID
-func ViewPlaybookRunFromChannelID(userID, channelID string, pluginAPI *pluginapi.Client) error {
+func ViewPlaybookRunFromChannelID(userID, channelID string, playbookService PlaybookService, playbookRunService PlaybookRunService, pluginAPI *pluginapi.Client) error {
```
Comment (lieut-data) [inline server/app/permissions.go:46]: nit: `UserCanViewPlaybookRunFromChannelID`?

File: server/app/permissions.go [hunk 5]
```diff
@@ -35,16 +52,12 @@ func ViewPlaybookRunFromChannelID(userID, channelID string, pluginAPI *pluginapi
...
-		return errors.Wrapf(err, "Unable to get channel to determine permissions, channel id `%s`", channelID)
-	}
-
-	if channel.Type == model.CHANNEL_OPEN && pluginAPI.User.HasPermissionToTeam(userID, channel.TeamId, model.PERMISSION_LIST_TEAM_CHANNELS) {
-		return nil
+		return errors.Wrapf(err, "Unable to get playbookRunID to determine permissions, channel id `%s`", channelID)
 	}
 
-	return ErrNoPermissions
+	return PlaybookAccess(userID, playbookRunID, playbookService, pluginAPI)
```
Comment (lieut-data) [inline server/app/permissions.go:60]: Passing `playbookRunID` but `PlaybookAccess` takes a `playbookID`?

File: server/sqlstore/playbook_run.go [hunk 6]
```diff
@@ -815,19 +815,18 @@ func (s *playbookRunStore) buildPermissionsExpr(info app.RequesterInfo) sq.Sqliz
...
-			  EXISTS(SELECT 1
-						 FROM ChannelMembers as cm
-						 WHERE cm.ChannelId = i.ChannelID
-						   AND cm.UserId = ?)
-			  -- Or if channel is public
-			  OR EXISTS(SELECT 1
-							FROM Channels as c
-							WHERE c.Id = i.ChannelID
-							  AND c.Type = 'O')
+ 			  NOT EXISTS(
```
Comment (lieut-data) [inline server/sqlstore/playbook_run.go:821]: Checking for the non-existence of a row to identify a positive permission "feels" weird. It seems accurate, but weird.

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks, @cpoile! A few comments, but also wondering if we can write an e2e test to exercise the permutations here to feel confident about access and such. The mocked unit tests remain unsatisfying for proving correctness :/

File: server/api/playbooks.go (hunk 1 above)
Comment (cpoile) [reply server/api/playbooks.go]: huh, that's weird, don't know how that got in there. Yep, gone. thanks! (Also cleaned up the other err2, err3s etc in the file)

File: server/app/permissions.go (hunk 3 above)
Comment (cpoile) [reply server/app/permissions.go]: Yep, good ideas. Done.

File: server/app/permissions.go (hunk 5 above)
Comment (cpoile) [reply server/app/permissions.go]: Oh man... That was a mistype. To be fair, `playbookRunID` and `playbookID` are too similar. 😝  You're right, the mocked tests don't do enough--I was also confusing `playbookID` and `playbookRunID` in the mocks! We need e2e tests for this: https://mattermost.atlassian.net/browse/MM-38136

File: server/sqlstore/playbook_run.go (hunk 6 above)
Comment (cpoile) [reply server/sqlstore/playbook_run.go:821]: That's how it's designed, sadly. An empty members list means the playbook is public. :). (ask Christopher. I think we all debated about it back then, too.)

**Comment by cpoile** [comment]
I've assigned myself https://mattermost.atlassian.net/browse/MM-38136 for next week so that I can work on the threaded replies.

**Review by lieut-data — APPROVED** [review]
:+1:

**Review by iomodo — APPROVED** [review]
Wonderful!

**Comment by cpoile** [comment]
@lieut-data I think this is ready to go now -- #719 verifies the different combinations work as expected. 

We'll need to make announcements on community explaining the new permissions scheme to everyone (hide your sensitive playbook names, checklists, and retros by making the playbook private). Maybe @justinegeffen should work on that, or is it okay if I do it?

**Comment by lieut-data** [comment]
@cpoile, please do take the lead on drafting and disseminating that information, but feel free to rely on @justinegeffen as needed!

**Comment by justinegeffen** [comment]
Happy for you to do it, @cpoile! And of course always happy to lend an editorial eye. :)

**Comment by cpoile** [comment]
@stephenvanhemmen This is an example notification. (Just a test for the checklists notification feature)

**Pull request merged**